netty/src/main/java/io/grpc/netty/NettyClientHandler.java

```diff
@@ -201,25 +200,42 @@ public void onStreamRemoved(Http2Stream stream) {
   @Override
   public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
           throws Exception {
-    if (msg instanceof CreateStreamCommand) {
-      createStream((CreateStreamCommand) msg, promise);
-    } else if (msg instanceof SendGrpcFrameCommand) {
-      sendGrpcFrame(ctx, (SendGrpcFrameCommand) msg, promise);
-    } else if (msg instanceof CancelClientStreamCommand) {
-      cancelStream(ctx, (CancelClientStreamCommand) msg, promise);
-    } else if (msg instanceof RequestMessagesCommand) {
-      ((RequestMessagesCommand) msg).requestMessages();
-    } else if (msg instanceof SendPingCommand) {
-      sendPingFrame(ctx, (SendPingCommand) msg, promise);
-    } else if (msg instanceof GracefulCloseCommand) {
-      gracefulClose(ctx, (GracefulCloseCommand) msg, promise);
-    } else if (msg instanceof ForcefulCloseCommand) {
-      forcefulClose(ctx, (ForcefulCloseCommand) msg, promise);
-    } else if (msg == NOOP_MESSAGE) {
-      ctx.write(Unpooled.EMPTY_BUFFER, promise);
-    } else {
-      throw new AssertionError("Write called for unexpected type: " + msg.getClass().getName());
+    UnionCommand cmd = (UnionCommand) msg;
+    switch (cmd.type()) {
+      case CANCEL_CLIENT_STREAM:
+        cancelStream(ctx, cmd, promise);
+        return;
+      case CANCEL_SERVER_STREAM:
+        break; // unsupported
+      case CREATE_STREAM:
+        createStream(cmd, promise);
+        return;
+      case FORCEFUL_CLOSE:
+        forcefulClose(ctx, cmd, promise);
+        return;
+      case GRACEFUL_CLOSE:
+        gracefulClose(ctx, cmd, promise);
+        return;
+      case REQUEST_MESSAGES:
+        cmd.requestMessagesCmdRequestMessages();
+        cmd.recycle();
+        return;
+      case SEND_GRPC_FRAME:
+        sendGrpcFrame(ctx, cmd, promise);
+        return;
+      case SEND_PING:
+        sendPingFrame(ctx, cmd, promise);
+        return;
+      case SEND_RESPONSE_HEADERS:
+        break; // unsupported
+      case NOOP:
+        ctx.write(Unpooled.EMPTY_BUFFER, promise);
+        cmd.recycle();
```

Comment:
Oh, I see. Each is being tracked separately. I think I'd prefer it being implicit and unconditional, because otherwise it's a pain to deal with in each method.